File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
  const n = friends.length;
  const idxByName = new Map();
  for (let i = 0; i < n; i++) idxByName.set(friends[i], i);

  // table[i][j] = i가 j에게 준 횟수
  const table = Array.from({ length: n }, () => Array(n).fill(0));
  // giftIndex[i] = (i가 준 총개수) - (i가 받은 총개수)
  const giftIndex = Array(n).fill(0);

  // 입력 처리
  for (const g of gifts) {
    const parts = g.trim().split(/\s+/);
    if (parts.length < 2) continue;
    const giver = parts[0], receiver = parts[1];
    if (!idxByName.has(giver) || !idxByName.has(receiver)) continue; // 안전장치
    const gi = idxByName.get(giver), ri = idxByName.get(receiver);
    table[gi][ri] += 1;
    giftIndex[gi] += 1;
    giftIndex[ri] -= 1;
  }

  // 다음 달 받을 예상 수
  const nextReceive = Array(n).fill(0);

  // 각 서로 다른 쌍(i, j) 에 대해 한 번만 판단 (i < j)
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const aToB = table[i][j];
      const bToA = table[j][i];
      if (aToB > bToA) {
        // i가 더 많이 줬으므로 i가 다음 달에 1개 받음
        nextReceive[i] += 1;
      } else if (aToB < bToA) {
        // j가 더 많이 줬으므로 j가 다음 달에 1개 받음
        nextReceive[j] += 1;
      } else {
        // 주고받은 수가 같거나 둘 다 0이면 giftIndex 비교
        if (giftIndex[i] > giftIndex[j]) nextReceive[i] += 1;
        else if (giftIndex[i] < giftIndex[j]) nextReceive[j] += 1;
        // 같으면 아무 변화 없음
      }
    }
  }

  // 가장 많이 받을 개수 반환
  return Math.max(...nextReceive);
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    print("친구", friends)
    print("선물", gifts)
     
    
    friends_dict = {} 
    for friend in friends:
        friends_dict[friend] = {}
        for f in friends:
            if friend == f:
                friends_dict[friend]['index'] = 0
            else:
                friends_dict[friend][f] = 0
    
    print(friends_dict)
    
    for gift in gifts:
        [giver, receiver] = gift.split()
        friends_dict[giver][receiver] += 1
        friends_dict[receiver][giver] -= 1
        
        friends_dict[giver]["index"] += 1
        friends_dict[receiver]["index"] -= 1
        
        
    print(friends_dict)
    max_gifts = 0
    
    for p, d in friends_dict.items():
        index = d['index']
        gift = 0
        for friend in friends:
            if friend != p:
                n = d[friend]
                if n > 0:
                    gift += 1
                elif n == 0:
                    if index > friends_dict[friend]['index']:
                        gift += 1 
                
        if gift > max_gifts:
            max_gifts = gift
    
    return max_gifts